File: bin/ModSecProcessor.pl
#!/usr/bin/perl
# Este script (cron) es el encargado de recoger los logs del mod_security y meterlos en la db para eso el mod_sec debe tener las siguientes directivas de configuracion:
#		SecAuditLogParts BFH
#		SecAuditLogType Concurrent
#		SecAuditLogStorageDir /LOGS/mod_sec/logs/
#
#
#
# La base de datos soportada es mysql y debe crearse un usuario con permisos SELECT,INSERT,UPDATE y DELETE.
# El create table es:
#	CREATE TABLE `log` (
#	  `id` int(11) NOT NULL AUTO_INCREMENT,
#	  `file_id` varchar(45) NOT NULL,
#	  `host` varchar(45) DEFAULT NULL,
#	  `get` varchar(255) DEFAULT NULL,
# 	  `date` varchar(45) DEFAULT NULL,
#	  `cookie` varchar(255) DEFAULT NULL,
#	  `referer` varchar(255) DEFAULT NULL,
#	  `ua` varchar(255) DEFAULT NULL,
#	  `xff` varchar(45) DEFAULT NULL,
#	  `ae` varchar(45) DEFAULT NULL,
#	  `message` text,
#	  `timestamp` varchar(20) DEFAULT NULL,
#	  PRIMARY KEY (`id`),
#	  UNIQUE KEY `file_id_UNIQUE` (`file_id`),
#	  KEY `timestamp` (`timestamp`),
#	  KEY `host` (`host`)
#	) ENGINE=InnoDB
#
#
#

use File::Find;
use File::Copy;
use Time::Local;
use DBI;

# Directorios a utilizar
my $tmpdir = "/tmp";
my $basedir = "/LOGS/mod_sec";
my $logsdir = $basedir . "/logs";
my $processed = $basedir . "/processed";

($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime(time);
$year += 1900;

my $sqlfile = "$sqlin" . "/" . "$year" . "$mon" . "$mday" . "$hour" . "$min" . ".sql";


sub parse_file{
	my $file = $_;

	if (-f $File::Find::name) {

		# Sacamos la fecha y la hora del ataque por el nombre del fichero y lo pasamos a timestamp para meterlo en la db
		$attack_year = substr $file, 0, 4; 
		$attack_month = substr $file, 4, 2; 
		$attack_month = $attack_month - 1;
		$attack_day = substr $file, 6, 2; 
		$attack_hour = substr $file, 9, 2;
		# Le sumamos 1 para poner la hora peninsular 
		#$attack_hour = $attack_hour + 1;
		if ($attack_hour == '24') { $attack_hour = 0;}
		$attack_min = substr $file, 11, 2; 
		$attack_sec = substr $file, 13, 2; 
		
		my $timestamp = timelocal($attack_sec,$attack_min,$attack_hour,$attack_day,$attack_month,$attack_year);
		

		open(FILE, $File::Find::name);
		while ( $line = <FILE>) {
			# Recogemos los varoles que nos interesan del log
			# GET 
			$get = $1 if ($line =~ /(GET.*)HTTP/);
			# Date
			$date = $1 if ($line =~ /(Date.*)/);
			# Host
			$host = $1 if ($line =~ /(Host.*)/);
			# Cookie 
			$cookie = $1 if ($line =~ /(Cookie.*)/);
			# Referer 
			$referer = $1 if ($line =~ /(Referer.*)/);
			# User-Agent 
			$ua = $1 if ($line =~ /(User-Agent.*)/);
			# X-Forwarded-For 
			$xff = $1 if ($line =~ /(X-Forwarded-For.*)/);
			# Accept-Encoding 
			$ae = $1 if ($line =~ /(Accept-Encoding.*)/);
			# Message 
			$message = $1 if ($line =~ /(Message.*)/);
		}
		close(FILE);

		my $query = "INSERT INTO log (file_id, host, get, date, cookie, referer, ua, xff, ae, message, timestamp) VALUES (" . $dbh->quote($file) . "," . $dbh->quote($host) . "," . $dbh->quote($get) . "," . $dbh->quote($date) . "," . $dbh->quote($cookie) . "," . $dbh->quote($referer) . "," . $dbh->quote($ua) . "," . $dbh->quote($xff) . "," . $dbh->quote($ae) . "," . $dbh->quote($message) . ","  . $dbh->quote($timestamp) . ");";

		$dbh->do($query);

		# Movemos los ficheros procesados al directorio processed
		move("$File::Find::name", "$processed" . "/" . "$file");
	
	}



}


# Conectamos con la DB
$dsn = "DBI:mysql:database=mod_sec;host=vipdbsistemas";
$dbh = DBI->connect($dsn, 'modsec', 'xxxxxxx');

# Hacemos un find de ficheros y los empezamos a procesar
find(\&parse_file, $logsdir);

$dbh->disconnect();

File: bin/ModSecProcessor.py
#! /opt/python/bin/python

import glob
from threading import *
import optparse
import os
import calendar
import time
import re
import MySQLdb as mdb

dbhost = 'localhost'
dbname = 'mod_sec'
dbuser = 'modsec'
dbpass = 'xxxxxx'

screenLock = Semaphore(value=1)

def InsertDB(query):
    # Con esta funcion conectamos a la db e insertamos
    global db
    
    try:
        screenLock.acquire()
        db = mdb.connect(dbhost,dbuser,dbpass,dbname)
        c = db.cursor()
        c.execute(query)
        db.commit()
        db.close()
        screenLock.release()
        return True
    except mdb.Error, e:
        print "\t\t\t[-] Database Error: " + str(e)
        print query
        screenLock.release()
        return False
    except:
        print "\t\t\t[-] Unknown Error"
        print query
        screenLock.release()
        return False



# Le pasamos como parametro el numero del hilo y fichero a procesar
def ProcessFiles(hilo, fichero):
    
    global get
    global host
    global date
    global referer
    global cookie
    global xff
    global ua
    global ae
    global message
    global l
    global file_content
    
    
    file_name = os.path.basename(fichero)
    attack_year = int(file_name[0:4])
    attack_month = int(file_name[4:6])
    attack_day = int(file_name[6:8])
    attack_hour = int(file_name[9:11])
    attack_min = int(file_name[11:13])
    attack_sec = int(file_name[13:15])
    
    
    t = (attack_year, attack_month, attack_day, attack_hour, attack_min, attack_sec)
    ts= calendar.timegm(t)
    # Le restamos una hora para que sea compatible con lo que busca la aplicacion web
    ts = ts - 3600
    
    
    # Abrimos el fichero para su parseo
    try:
        print "\t\t[+] " + str(hilo) + ":" + fichero
        f = open(fichero, 'r')
        file_content = str(f.readlines(1000000))
        f.close()
    except:
        print "\t\t\t[-] " + fichero + ": No Existe!"
        return False
        
    
    l = re.sub('\'','', file_content)
    lines = []
    lines = l.split('\\n')
    get = ""
    host = ""
    date = ""
    cookie = ""
    referer = ""
    ua = ""
    ae = ""
    xff = ""
    message = ""
    
    for line in lines:
        line = re.sub('^,','', line)
        
        
        
        m = re.search('([GET|POST].*)HTTP', line)
        if m:
            get = m.group(1)

        m = re.search('(Date.*)', line)
        if m:
            date = m.group(1)

        m = re.search('(Host.*)', line)
        if m:
            host = m.group(1)

        m = re.search('(Cookie.*)', line)
        if m:
            cookie = m.group(1)

        m = re.search('(Referer.*)', line)
        if m:
            referer = m.group(1)

        m = re.search('(User-Agent.*)', line)
        if m:
            ua = m.group(1)

        m = re.search('(Accept-Encoding.*)', line)
        if m:
            ae = m.group(1)

        m = re.search('(X-Forwarded-For.*)', line)
        if m:
            xff = m.group(1)

        m = re.search('(Message.*)', line)
        if m:
            message = m.group(1)
       
    
     
    
    query = "INSERT INTO log (file_id, host, get, date, cookie, referer, ua, xff, ae, message, timestamp) VALUES (\'" + mdb.escape_string(file_name) + "\', \'" + mdb.escape_string(host) + "\', \'" + mdb.escape_string(get) + "\', \'" + mdb.escape_string(date) + "\', \'" + mdb.escape_string(cookie) + "\', \'" + mdb.escape_string(referer) + "\', \'" + mdb.escape_string(ua) + "\', \'" + mdb.escape_string(xff) + "\', \'" + mdb.escape_string(ae) + "\', \'" + mdb.escape_string(message) + "\', \'" + str(ts) + "\');"

    InsertDB(query)
    #print "\t\t\t[-] Borrando: " + fichero
    os.remove(fichero)
    
def main():
    
    basedir = "/LOGS/mod_sec"
    logsdir = basedir + "/logs"
    global files
    
    parser = optparse.OptionParser('./ModSecProcessor.py ' + '-t <threads>')
    parser.add_option('-t', dest='concurrent', type='int',help='Especifica el numero de threads a utilizar para lanzar el proceso')
        
    (options, args) = parser.parse_args()
        
    concurrent = options.concurrent
            
    if (concurrent == None):
        print parser.usage
        exit(0)
    
    print " [*] Obtiendo listado de ficheros a procesar:"
    # Formato del fichero a buscar: 20130213-104032-URtfkKwSAxYAAHYN9-IAAAA 
    
    files = glob.glob(logsdir + '*/*/*/*');
    toprocess = len(files)
    print "\t [+] Ficheros a procesar " + str(toprocess)
        
        
    count = 0    
    # Lanzamos los threads
    for i in range(toprocess):
        count += 1
        t = Thread(target=ProcessFiles, args=(i,files[i]))
        t.start()
        if count == concurrent:
            t.join()
            count = 0
        
    
if __name__ == "__main__":
    main()
